File: src/com/gnarly/game/Snake.java
package com.gnarly.game;

import static org.lwjgl.glfw.GLFW.GLFW_KEY_A;
import static org.lwjgl.glfw.GLFW.GLFW_KEY_D;
import static org.lwjgl.glfw.GLFW.GLFW_KEY_DOWN;
import static org.lwjgl.glfw.GLFW.GLFW_KEY_LEFT;
import static org.lwjgl.glfw.GLFW.GLFW_KEY_RIGHT;
import static org.lwjgl.glfw.GLFW.GLFW_KEY_S;
import static org.lwjgl.glfw.GLFW.GLFW_KEY_UP;
import static org.lwjgl.glfw.GLFW.GLFW_KEY_W;

import java.util.LinkedList;

import org.joml.Vector2f;

import com.gnarly.engine.display.Camera;
import com.gnarly.engine.display.Window;
import com.gnarly.engine.model.ColRect;

public class Snake {

	public static final Vector2f
		DIR_UP    = new Vector2f( 0, -1),
		DIR_DOWN  = new Vector2f( 0,  1),
		DIR_LEFT  = new Vector2f(-1,  0),
		DIR_RIGHT = new Vector2f( 1,  0);
	
	public static final int
		DIR_UP_AR    = 0,
		DIR_DOWN_AR  = 1,
		DIR_LEFT_AR  = 2,
		DIR_RIGHT_AR = 3;
	
	public static final Vector2f[]
		DIRS = { DIR_UP, DIR_DOWN, DIR_LEFT, DIR_RIGHT };
	
	private static ColRect rect = null;
	
	private Window window;
	private Camera camera;
	
	private Vector2f dir, banned = null, future;
	private LinkedList<Vector2f> snake = new LinkedList<Vector2f>();
	
	private int addLength = 0;
	
	public Snake(Window window, Camera camera) {
		this.window = window;
		this.camera = camera;
		if(rect == null)
			rect = new ColRect(camera, 0, 0, -0.08f, 1, 1, 0, 1, 0, 1, false);
	}
	
	public void update(boolean full) {
		if (full) {
			snake.add(0, snake.get(0).add(future, new Vector2f()));
			if(addLength == 0)
				snake.removeLast();
			else
				--addLength;
			future = dir;
			banned = future;
		}
		if(window.keyPressed(GLFW_KEY_LEFT) > 0 || window.keyPressed(GLFW_KEY_A) > 0 && banned != DIR_RIGHT)
			dir = DIR_LEFT;
		else if(window.keyPressed(GLFW_KEY_RIGHT) > 0 || window.keyPressed(GLFW_KEY_D) > 0 && banned != DIR_LEFT)
			dir = DIR_RIGHT;
		else if(window.keyPressed(GLFW_KEY_UP) > 0 || window.keyPressed(GLFW_KEY_W) > 0 && banned != DIR_DOWN)
			dir = DIR_UP;
		else if(window.keyPressed(GLFW_KEY_DOWN) > 0 || window.keyPressed(GLFW_KEY_S) > 0 && banned != DIR_UP)
			dir = DIR_DOWN;
	}
	
	public boolean check(int width, int height) {
		Vector2f head = snake.get(0);
		if (head.x < 0 || head.x == width || head.y < 0 || head.y == height)
			return false;
		for (int i = 1; i < snake.size(); i++)
			if (head.equals(snake.get(i)))
				return false;
		return true;
	}
	
	public void lengthen() {
		++addLength;
	}
	
	public void lengthen(int length) {
		addLength += length;
	}
	
	public Vector2f getHead() {
		return new Vector2f(snake.get(0));
	}
	
	public Vector2f getFuture() {
		return snake.get(0).add(future, new Vector2f());
	}
	
	public Vector2f getFuture(float percent) {
		return new Vector2f(snake.get(0).add(future.mul(percent, new Vector2f()), new Vector2f()));
	}
	
	public boolean contains(Vector2f position) {
		for (int i = 0; i < snake.size(); i++)
			if(snake.get(i).equals(position))
				return true;
		return false;
	}
	
	public void render(float percent) {
		if(addLength > 0) {
			rect.setPosition(snake.getLast());
			rect.render();
		}
		snake.add(0, snake.get(0).add(future, new Vector2f()));
		for (int i = 1; i < snake.size(); i++) {
			rect.setPosition(snake.get(i).add(snake.get(i - 1).sub(snake.get(i), new Vector2f()).mul(percent), new Vector2f()));
			rect.render();
			if(i != snake.size() - 1 && snake.get(i - 1).x != snake.get(i + 1).x && snake.get(i - 1).y != snake.get(i + 1).y) {
				rect.setPosition(snake.get(i));
				rect.render();
			}
		}
		snake.remove(0);
	}
	
	public void reset(int x, int y, Vector2f dir) {
		snake.clear();
		snake.add(new Vector2f(x, y));
		banned = dir;
		this.dir = dir;
		future = dir;
	}
}

File: src/com/gnarly/game/Snake.kt
package com.gnarly.game

import com.gnarly.engine.Camera
import com.gnarly.engine.Vector
import com.gnarly.engine.Window
import org.joml.Matrix4f
import org.lwjgl.glfw.GLFW.*

class Snake(
	var futureDirection: Direction,
    var lengthToAdd: Int,
    initialPoint: Point
) {
	class InputBuffer() {
		var primary: Direction? = null
		var secondary: Direction? = null

		fun pop(): Direction? {
			if (primary != null) {
				val ret = primary
				primary = null
				return ret
			} else if (secondary != null) {
				val ret = secondary
				secondary = null
				return ret
			}

			return null
		}
	}

	var fallbackPoint = initialPoint

	data class Segment(var direction: Direction, var point: Point)

	private val inputBuffer = InputBuffer()
	private val snake: ArrayList<Segment> = arrayListOf(Segment(futureDirection, initialPoint))

	fun update(window: Window, allowInput: Boolean, addSegment: Boolean, map: Map) {
		if (addSegment && snake.isNotEmpty()) {
			if (lengthToAdd >= 0) {
				val newDirection = inputBuffer.pop() ?: futureDirection
				val newPoint = snake.first().point + futureDirection.point

				snake.add(0, Segment(newDirection, newPoint))

				fallbackPoint = newPoint
				futureDirection = newDirection

			} else {
				++lengthToAdd
			}

			if (lengthToAdd <= 0) {
				snake.removeLastOrNull()
			} else {
				--lengthToAdd
			}
		}

		if (allowInput) {
			val inputtedDirections = arrayOf(
				window.key(GLFW_KEY_RIGHT) >= GLFW_PRESS || window.key(GLFW_KEY_D) >= GLFW_PRESS,
				window.key(GLFW_KEY_UP) >= GLFW_PRESS || window.key(GLFW_KEY_W) >= GLFW_PRESS,
				window.key(GLFW_KEY_LEFT) >= GLFW_PRESS || window.key(GLFW_KEY_A) >= GLFW_PRESS,
				window.key(GLFW_KEY_DOWN) >= GLFW_PRESS || window.key(GLFW_KEY_S) >= GLFW_PRESS
			)

			val primaryLeft = futureDirection.rotateLeft()
			val primaryRight = futureDirection.rotateRight()

			/* trying to turn in both directions at the same time */
			if (inputtedDirections[primaryLeft.ordinal] && inputtedDirections[primaryRight.ordinal]) {
				return
			}

			val primary = if (inputtedDirections[primaryLeft.ordinal]) primaryLeft
				else if (inputtedDirections[primaryRight.ordinal]) primaryRight
				else null

			if (inputBuffer.primary == null) {
				if (primary != null) {
					val wallCheckPos = snake.first().point + futureDirection.point + primary.point
					if (map.access(wallCheckPos.x, wallCheckPos.y) != MapTemplate.TYPE_WALL) {
						inputBuffer.primary = primary
					}
				}
			} else {
				primary?.let { inputBuffer.primary = it }
			}
		}
	}

	fun bodyContains(point: Point): Boolean {
		for (i in 1 until snake.size) if (point == snake[i].point) return true
		return false
	}

	fun lengthen(addedLength: Int) {
		lengthToAdd += addedLength
	}

	fun reduceToNothing() {
		lengthToAdd = -snake.size
	}

	fun getCameraPos(percent: Float): Vector {
		return if (lengthToAdd < 0 || snake.isEmpty()) {
			(snake.firstOrNull()?.point ?: fallbackPoint).toVector() + Vector(0.5f, 0.5f)
		} else {
			(snake.firstOrNull()?.point ?: fallbackPoint).toVector() + Vector(0.5f, 0.5f) + (futureDirection.point.toVector() * percent)
		}
	}

	fun getOn() = snake.first().point
	fun getMovingInto() = snake.first().point + futureDirection.point

	private fun Camera.snakeEndModel(x: Float, y: Float, width: Float, height: Float, rotation: Float): Matrix4f {
		return model.translation(x, y, 0.0f).translate(0.5f, 0.5f, 0.0f).rotateZ(rotation).translate(-0.5f, -0.5f, 0.0f).scale(width, height, 0f)
	}

	fun render(camera: Camera, percent: Float) {
		Assets.colorShader.enable()
		Assets.colorShader.setColor(0.0f, 1.0f, 0.0f, 1.0f)

		/* render future */
		if (lengthToAdd >= 0 && snake.isNotEmpty()) {
			val futurePoint = snake.first().point + futureDirection.point

			Assets.colorShader.setMVP(
				camera.projectionView(),
				camera.snakeEndModel(
					futurePoint.x.toFloat(),
					futurePoint.y.toFloat(),
					percent,
					1.0f,
					futureDirection.rotation
				)
			)
			Assets.rect.render()
		}

		/* render body segments */
		for (i in 0..snake.lastIndex - 1) {
			val segment = snake[i]
			Assets.colorShader.setMVP(
				camera.projectionView(),
				camera.model(segment.point.x.toFloat(), segment.point.y.toFloat(), 1.0f, 1.0f)
			)
			Assets.rect.render()
		}

		/* render tail */
		snake.lastOrNull()?.let { (tailDirection, tailPoint) ->
			if (lengthToAdd > 0) {
				Assets.colorShader.setMVP(camera.projectionView(), camera.model(tailPoint.x.toFloat(), tailPoint.y.toFloat(), 1.0f, 1.0f))
			} else {
				Assets.colorShader.setMVP(camera.projectionView(), camera.snakeEndModel(tailPoint.x.toFloat(), tailPoint.y.toFloat(), 1.0f - percent, 1.0f, tailDirection.inverse().rotation))
			}
			Assets.rect.render()
		}
	}
}
